### PR TITLE
Convert `indexes.py` to use `Self` for typing

### PR DESCRIPTION
**Pull request by max-sixty**
I wanted to do this separately from #8216 , as it's the only one that adds a couple `ignore`s. And given the ratio of impact-to-potential-merge-conflicts, I didn't want to slow the other PR down.

### CODE REVIEW

File: xarray/core/indexes.py [hunk 1]
```diff
@@ -304,7 +302,7 @@ def reindex_like(self: T_Index, other: T_Index) -> dict[Hashable, Any]:
         """
         raise NotImplementedError(f"{self!r} doesn't support re-indexing labels")
 
-    def equals(self: T_Index, other: T_Index) -> bool:
+    def equals(self, other: T_Index) -> bool:
```
Comment (headtr1ck) [inline xarray/core/indexes.py:305]:
```suggestion
    def equals(self, other: Self) -> bool:
```

File: xarray/core/indexes.py [hunk 2]
```diff
@@ -674,10 +670,11 @@ def _concat_indexes(indexes, dim, positions=None) -> pd.Index:
     @classmethod
     def concat(
         cls,
-        indexes: Sequence[PandasIndex],
+        # Currently only compatible with PandasIndex, which fails the Liskov substitution principle
+        indexes: Sequence[Self],  # type: ignore[override]
```
Comment (headtr1ck) [inline xarray/core/indexes.py:674]: If you change the type of indexes to `Sequence[Self]` in `Index.concat` as well, does this work?
Or is it supposed to work with any indexes in the future?

File: xarray/core/indexes.py [hunk 3]
```diff
@@ -264,7 +262,7 @@ def sel(self, labels: dict[Any, Any]) -> IndexSelResult:
         """
         raise NotImplementedError(f"{self!r} doesn't support label-based selection")
 
-    def join(self: T_Index, other: T_Index, how: JoinOptions = "inner") -> T_Index:
+    def join(self, other: T_Index, how: JoinOptions = "inner") -> Self:
```
Comment (headtr1ck) [inline xarray/core/indexes.py:265]:
```suggestion
    def join(self, other: Self, how: JoinOptions = "inner") -> Self:
```
Not sure if this is supposed to support any indexes again.

File: xarray/core/indexes.py [hunk 4]
```diff
@@ -286,7 +284,7 @@ def join(self: T_Index, other: T_Index, how: JoinOptions = "inner") -> T_Index:
             f"{self!r} doesn't support alignment with inner/outer join method"
         )
 
-    def reindex_like(self: T_Index, other: T_Index) -> dict[Hashable, Any]:
+    def reindex_like(self, other: T_Index) -> dict[Hashable, Any]:
```
Comment (headtr1ck) [inline xarray/core/indexes.py:287]:
```suggestion
    def reindex_like(self, other: Self) -> dict[Hashable, Any]:
```
Same thing here.

**Review by headtr1ck — COMMENTED** [review]
Your title says Variable but you have changed the index module ;)

Just a couple minor suggestions

File: xarray/core/indexes.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/indexes.py:674]: Very good idea! This removes lots of ignores

File: xarray/core/indexes.py [hunk 5]
```diff
@@ -811,7 +811,10 @@ def join(self: PandasIndex, other: PandasIndex, how: str = "inner") -> PandasInd
         return type(self)(index, self.dim, coord_dtype=coord_dtype)
 
     def reindex_like(
-        self, other: PandasIndex, method=None, tolerance=None
+        self,
+        other: Self,
+        method=None,
+        tolerance=None,
```
Comment (headtr1ck) [inline xarray/core/indexes.py:817]:
```suggestion
        tolerance=None
```

**Review by headtr1ck — APPROVED** [review]

**Pull request merged**